### PR TITLE
Add Typescript implementation of the SDK

### PR DESCRIPTION
**Pull request by kenrick**
## Overview

This PR aims to add the javascript implementation of the Prelude SDK. This will be closely based on the python version and will be done in Typescript. The package will be called `@prelude/sdk` on NPM.

### Changes and TODOs

- [x] add Build controller methods
   - [x] Add types and interfaces for the responses from the API
- [x] add IAM controller methods
- [x] add Detect controller methods

### CODE REVIEW

**Comment by kenrick** [comment]
Questions;
1. What is the relation ship between TTP ids and their code file names? -- I tired to add a code with the name `linux-{uuid}-x86.c` to a TTP with that {uuid} as its id and got an error about the name being incorrect

Issues while testing:
1.  I tried to run the `deleteTTP` and `deleteCodeFile` against https://detect.dev.prelude.org but doesn't seem to work. 
2. CORS not set to allow requests directly from browsers

cc @FM @privateducky

File: javascript/sdk/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+## Install
+
+```bash
+npm install @prelude/sdk
+```
+
+## Quick start
+
+The following example registers an account and gets manifest:
```
Comment (privateducky) [inline javascript/sdk/README.md:13]: gets the manifest

File: javascript/sdk/README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+```typescript
+import * as Prelude from "@prelude/sdk";
+const service = new Prelude.Service({ host: "https://detect.dev.prelude.org" });
```
Comment (privateducky) [inline javascript/sdk/README.md:17]: update to prod url (drop .dev)

File: javascript/sdk/index.html [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      const manifests = await service.build.listManifest();
+
+      console.log("Manifests before create", manifests);
+
+      const ttpId = uuid.v4();
+      await service.build.createTTP(ttpId, "This is a test");
+
+      const manifests2 = await service.build.listManifest();
+
+      console.log("Manifests after create", manifests2);
```
Comment (privateducky) [inline javascript/sdk/index.html:30]: anywhere you use 'manifests' swap to 'manifest' for proper usage

File: javascript/sdk/lib/controllers/build.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
+import Client from "../client";
```
Comment (privateducky) [inline javascript/sdk/lib/controllers/build.ts:1]: I added a few 'purge' routes on the API that we'll need on the FE; might as well add here anytime.

File: javascript/sdk/lib/controllers/detect.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      ...options,
+    });
+
+    return response.text();
+  }
+
+  async endpointActivity(
+    { endpointId, days = 7 }: { endpointId: string; days?: number },
+    options: RequestOptions
+  ) {
```
Comment (privateducky) [inline javascript/sdk/lib/controllers/detect.ts:33]: this route has been deprecated; can remove

File: javascript/sdk/lib/types.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  MONTHLY: 2,
+  ONCE: 3,
+} as const;
+
+export type RunCode = typeof RunCodes[keyof typeof RunCodes];
+
+export const Permissions = {
+  ADMIN: 0,
+  EXECUTIVE: 1,
+  SERVICE: 2,
```
Comment (privateducky) [inline javascript/sdk/lib/types.ts:46]: swap service/operator permissions - this was a quick change recently

File: javascript/sdk/index.html (hunk 3 above)
Comment (kenrick) [reply javascript/sdk/index.html:30]: Do you get the manifest that was just created?

File: javascript/sdk/lib/types.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+export interface EndpointActivity {}
+
+export interface AccountActivity {}
+
+export interface AccountQueue {}
+
+export const RunCodes = {
+  DEBUG: 0,
+  DAILY: 1,
+  MONTHLY: 2,
```
Comment (privateducky) [inline javascript/sdk/lib/types.ts:37]: make this:
```
DEBUG: 0
DAILY: 1
WEEKLY: 2
MONTHLY: 3
ONCE: 4
```

(recent update)

File: javascript/sdk/yarn.lock [hunk 8]
```diff
@@ -0,0 +1,242 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (privateducky) [inline javascript/sdk/yarn.lock:1]: remove + add to .gitignore

File: javascript/sdk/README.md [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+```typescript
+import * as Prelude from "@prelude/sdk";
+const service = new Prelude.Service({ host: "https://detect.prelude.org" });
+
+const credentials = await service.iam.newAccount("test@example.com");
+
+service.setCredentials(credentials);
+
+const manifests = await service.build.listManifest();
```
Comment (privateducky) [inline javascript/sdk/README.md:23]: swap 'manifests' to 'manifest' everywhere (weird word in general with the plural state of it - blame my journalism background!)

**Review by privateducky — CHANGES_REQUESTED** [review]
left a handful of minor adjustments, mostly recent changes that you wouldn't have caught. otherwise - fantastic work!

File: javascript/sdk/index.html (hunk 3 above)
Comment (kenrick) [reply javascript/sdk/index.html:30]: Got it

File: javascript/sdk/README.md (hunk 9 above)
Comment (kenrick) [reply javascript/sdk/README.md:23]: lol

File: javascript/sdk/yarn.lock (hunk 8 above)
Comment (kenrick) [reply javascript/sdk/yarn.lock]: Would love to know your reasoning as to why later on

**Review by privateducky — APPROVED** [review]

**Pull request merged**